docs/proposals/ONNXMultiDeviceProposal.md

```diff
@@ -0,0 +1,181 @@
+<!--
+Copyright (c) ONNX Project Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+# ONNX Multi-Device Proposal
+
+## Background
+
+The recent trend in increasingly larger models has spurred an interest in distributed inference. A key performance bottleneck for inference for these large models has been the memory limits of GPUs and other accelerators as well as communication bandwidth. Thus, efficient distributed inference typically requires parallelization of the computation across multiple devices taking memory and bandwidth into account.
+
+Our goal is to extend ONNX so that it can serve as a representation of a parallelized model. This is driven by the current state-of-the-art techniques used for distributed inference (eg., see [GSPMD: General and Scalable Parallelization for ML Computation Graphs](https://arxiv.org/pdf/2105.04663.pdf)). In particular, two techniques of interest are tensor parallelism and pipelining. In tensor parallelism (also known as horizontal parallelism or operator parallelism), the computation of a single operator (node) in the graph is parallelized across multiple devices by sharding its inputs, In pipeline parallelism, different subgraphs are assigned to different devices.
+
+
+## Design
+
+See [this commit](https://github.com/kevinch-nv/onnx/commit/07e97452096b28ba7c46fec6927d195907431e07) for the proposed additions to the ONNX spec.
+
+The key point of this design is that all multi-device specific annotations are at the node level, and do not affect the main computational graph. This means:
+ - All communication operations required for multi-device execution are implicit
+ - A backend may choose to ignore the annotations if the provided configurations are either not supported or not available
+
+### Sharding Specification
```

Comment:
@kevinch-nv can we ensure that the ONNX spec aligns with XLA's new proposed sharding spec: https://github.com/openxla/shardy/blob/main/docs/sharding_representation.md

Shardy claims to have also learned from other implementations from which we could benefit. Details can be found [here](https://llvm.org/devmtg/2024-10/slides/techtalk/Chrzaszcz-Jiang-Shardy.pdf).
![image](https://github.com/user-attachments/assets/233d83c9-aa61-417c-bfe5-7926e604c58e)

Reply:
Thanks for the pointer. Let me read through the spec and see if ONNX can adopt something similar.

Reply:
We're going to keep the spec as-is for this initial proposal.